swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift

```diff
@@ -109,7 +110,40 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
 
   func handleTunnelShutdown(dueTo reason: TunnelShutdownEvent.Reason, errorMessage: String) {
     TunnelShutdownEvent.saveToDisk(reason: reason, errorMessage: errorMessage, logger: self.logger)
+
+    #if os(iOS)
+      if reason.action == .signoutImmediately {
+        UNUserNotificationCenter.current().getNotificationSettings { notificationSettings in
+          if notificationSettings.authorizationStatus == .authorized {
+            self.logger.log(
+              "Notifications are allowed. Alert style is \(notificationSettings.alertStyle.rawValue)"
+            )
+            self.showSignedOutNotificationiOS()
+          }
+        }
+      }
+    #endif
   }
+
+  #if os(iOS)
+    func showSignedOutNotificationiOS() {
+      let content = UNMutableNotificationContent()
+      content.title = "Your Firezone session has ended"
+      content.body = "Please sign in again to reconnect"
+      content.categoryIdentifier = NotificationIndentifier.sessionEndedNotificationCategory.rawValue
+      let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 1, repeats: false)
+      let request = UNNotificationRequest(
+        identifier: "FirezoneTunnelShutdown", content: content, trigger: trigger
+      )
+      UNUserNotificationCenter.current().add(request) { error in
+        if let error = error {
+          self.logger.error("showSignedOutNotificationiOS: Error requesting notification: \(error)")
+        } else {
+          self.logger.error("showSignedOutNotificationiOS: Successfully requested notification")
+        }
+      }
+    }
+  #endif
```

Comment:
Same here -- you could extract this into a single `SessionNotification.showSignedOutNotification()` function and in the `SessionNotification` class have the macOS / iOS conditional.

Reply:
👍